fvm/programs/chain_programs.go

```diff
@@ -0,0 +1,104 @@
+package programs
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const DefaultProgramsCacheSize = 1000
+
+// ChainPrograms is a cache of BlockPrograms databases used for speeding up
+// cadence execution.
+//
+// Since programs are derived from external source, the BlockPrograms databases
+// need not be durable and can be recreated on the fly.
+type ChainPrograms struct {
+	mutex sync.Mutex
+
+	lru *simplelru.LRU
+}
+
+func NewChainPrograms(chainCacheSize uint) (*ChainPrograms, error) {
+	lru, err := simplelru.NewLRU(int(chainCacheSize), nil)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create LRU cache: %w", err)
+	}
+
+	return &ChainPrograms{
+		lru: lru,
+	}, nil
+}
+
+func (chain *ChainPrograms) Get(
+	currentBlockId flow.Identifier,
+) *Programs {
+	chain.mutex.Lock()
+	defer chain.mutex.Unlock()
+
+	currentEntry, ok := chain.lru.Get(currentBlockId)
+	if ok {
+		return currentEntry.(*Programs)
+	}
+
+	return nil
+}
+
+func (chain *ChainPrograms) GetOrCreateBlockPrograms(
+	currentBlockId flow.Identifier,
+	parentBlockId flow.Identifier,
+) *Programs {
+	chain.mutex.Lock()
+	defer chain.mutex.Unlock()
+
+	currentEntry, ok := chain.lru.Get(currentBlockId)
+	if ok {
+		return currentEntry.(*Programs)
+	}
+
+	var current *Programs
+	parentEntry, ok := chain.lru.Get(parentBlockId)
+	if ok {
+		current = parentEntry.(*Programs).ChildPrograms()
+	} else {
+		current = NewEmptyPrograms()
+	}
+
+	chain.lru.Add(currentBlockId, current)
+	return current
+}
+
+func (chain *ChainPrograms) getCurrentOrParent(
+	currentBlockId flow.Identifier,
+	parentBlockId flow.Identifier,
+) *Programs {
+	chain.mutex.Lock()
+	defer chain.mutex.Unlock()
+
+	entry, ok := chain.lru.Get(currentBlockId)
+	if ok {
+		return entry.(*Programs)
+	}
+
+	entry, ok = chain.lru.Get(parentBlockId)
+	if ok {
+		return entry.(*Programs)
+	}
+
+	return nil
+}
+
+func (chain *ChainPrograms) NewBlockProgramsForScript(
+	currentBlockId flow.Identifier,
+	parentBlockId flow.Identifier,
```

Comment:
nit: i assume the invariant here is that `parentBlockId` should be actually the id of `currentBlockId` 's parent node. But if we expose these params, there's a slight chance that caller may mistakenly call it with this invariant broken. not blocking to me, just sth to think about.